python/ray/serve/controller.py

```diff
@@ -255,6 +251,10 @@ async def run_control_loop(self) -> None:
                     self.deployment_state_manager.update()
                 except Exception:
                     logger.exception("Exception updating deployment state.")
+                try:
+                    self.application_state_manager.update()
```

Comment:
The [`ApplicationStateManager`](https://github.com/ray-project/ray/pull/31589/files#diff-31975e75fad7092f05d8d5759dfeb90d8b454fc4193e4970b55914dbb87ad968R142) feels like an awkward abstraction imo. On every control loop, the `ServeController` will update the deployments by calling the `deployment_state_manager` directly, and then it'll again update the deployments by calling the `deployment_state_manager` indirectly via the `application_state_manager`.

This makes it harder to understand how deployments are changed in a single control loop because there's two places where they're being updated. This also makes it harder to reason about concurrency (e.g. what happens if two apps have deployments w/ the same name? How would the `deployment_state_manager`/`application_state_manager` handle that?).

Since an application is essentially a collection of deployments and an application name, could we instead make it as a dictionary in the `DeploymentStateManager`? Status calls could then be made directly to the `DeploymentStateManager`, which would produce a result based on the dictionary.

Reply:
1. Currently there is no update call from `application_state_manager` to `deployment_state_manager`, application manager only fetches information from the deployment state manager. Since we are going to deprecate the v1 API. It make sense to me we directly hide `deployment_state_manager` into `application_state_manager`, so which means we only call `application_state_manager.update()` in the future. (I can't do it for now because we still support v1 API)
2. For `what happens if two apps have deployments w/ the same name`, it will deploy the one that comes later.
3. `make it as a dictionary in the DeploymentStateManager`.  One reason I don't do that is that we still need to have the abstraction for application level even in the DeploymentStateManager.  DeploymentStateManager class itself is used for deploying single Deployment, it doesn't have group of Deployments concept, and no application concept. we have to introduce it somehow. Given my first point, So I decide to have another class to reflect the application level information. And again after v1 deprecation, we can hide `deployment_state_manager` into the `application_state_manager`.

Reply:
I see, so eventually the `deployment_state_manager` will only get called via the `application_state_manager`. I think that makes sense.

> For `what happens if two apps have deployments w/ the same name`, it will deploy the one that comes later.

Does that mean two apps on a single cluster can't have deployments with the same name at the same time?

Reply:
Yes, we will have application_state_manager.update() only in the controller level.

I think you are referring to the problem here https://github.com/ray-project/ray/pull/31589#issuecomment-1387446728 :)
I am working on this so basically each deployment name will be `{app_name}_{deployment_name}`. It will reflect in the following up update in this pr soon!

Reply:
@sihanwang41 I thought that currently, `DeploymentStateManager` is used for deploying multiple deployments? It holds a [list](https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/deployment_state.py#L1808) of `DeploymentStates`. Essentially from my understanding it should manage the state of a single application.

Reply:
@zcin yes, it is good for single applications, if for multiple applications, we have to introducing `group deployments` and `app concept` somewhere.

Reply:
Yeah seems like this PR is the lowest friction path to add this concept, but it would probably make more sense to refactor and have "app manager" wrap "deployment manager" in the future.